nav2_collision_monitor/src/pointcloud.cpp

```diff
@@ -65,17 +65,24 @@ void PointCloud::configure()
     std::bind(&PointCloud::dataCallback, this, std::placeholders::_1));
 }
 
-void PointCloud::getData(
+bool PointCloud::getData(
   const rclcpp::Time & curr_time,
   std::vector<Point> & data) const
 {
   // Ignore data from the source if it is not being published yet or
   // not published for a long time
   if (data_ == nullptr) {
-    return;
+    if (source_timeout_.seconds() == 0.0) {
+      return true;
+    }
+    return false;
   }
   if (!sourceValid(data_->header.stamp, curr_time)) {
-    return;
+    // Don't block if source_timeout == 0.0
+    if (source_timeout_.seconds() == 0.0) {
+      return true;
+    }
+    return false;
```

Comment:
It looks like we have to face the same as reported in https://github.com/ros-planning/navigation2/pull/3880#discussion_r1361601175 issue.

Let's check the case when `source_timout_` is equal to `0.0`: this is the case, when we should ignore outdated source. However, how could we determine that the source is outdated?
In other words, during the execution `sourceValid()` routine contains the following comparison inside:
```
 const rclcpp::Duration dt = curr_time - source_time;
  if (dt > source_timeout_) {
    // Fail-case
  }
```
which always will return `false`. Thus, if we set 0-value of `source_timeout` for specific source, this source will be always ignored, independently on which timeout we want to bring on, after the ignoring became to be.

Reply:
Gosh. I get it know. Nice catch. This is problematic, a refactor is needed. I don't see a way to solve it without re-introducing the additional bool parameter `block_if_invalid` (or better name). Thoughts @SteveMacenski ?

Reply:
@doisyg why not just move this check inside of `sourceValid`? 

```

if (source_timeout_.seconds() != 0.0 && dt > source_timeout_) {
```

Reply:
That will push the problem down to the next part of `getData` and we will have a tf error at some point :
https://github.com/doisyg/navigation2/blob/e54293f9b327174de619519317b6bb5bc0d2e282/nav2_collision_monitor/src/pointcloud.cpp#L89-L111

Though if we also don't block on tf error and just skip the data on `source_timeout == 0.0`, we could add a check higher here (which will require a new source::getSourceTimeout method) and we should be fine: https://github.com/doisyg/navigation2/blob/e54293f9b327174de619519317b6bb5bc0d2e282/nav2_collision_monitor/src/collision_monitor_node.cpp#L389)

```
  // Fill collision_points array from different data sources
  for (std::shared_ptr<Source> source : sources_) {
    if (source->getEnabled()) {
      if (!source->getData(curr_time, collision_points) && source->getSourceTimeout() != 0.0) {
        action_polygon = nullptr;
        robot_action.polygon_name = "invalid source";
        robot_action.action_type = STOP;
        robot_action.req_vel.x = 0.0;
        robot_action.req_vel.y = 0.0;
        robot_action.req_vel.tw = 0.0;
        break;
      }
    }
  }

```
Pushing this now

Reply:
> we will have a tf error at some point 

So thinking aloud: 
- If we check if source is valid `sourceValid` with the check on timeout, then we either return `false` due to stale data with a timeout or...
- We continue on allowing stale data without a timeout. Then:
- We try to TF transform it. As long as that data isn't insanely stale, that should be fine as long as its still in the TF buffer
- If its not in the TF buffer, then the `return false` is warranted since its literally impossible to use 

That seems like a fine workflow. If we set no timeout, then it should work fine for as long as the data is in any way possibly actionable. 

So then going to the `node` code block you added, that would make sense to not `STOP` in that case. We don't want to consider the data, but we also don't want to stop either. That seems sensible.